rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -44,33 +62,70 @@
  */
 public class FlowNodeUtil {
 
+    static class ExecutionCache {
+        List<FlowNode> idSortedExecutionNodeList;
+    }
+
+    // Stores ExecutionInfo
+    static final Cache<String,ExecutionCache> executionCache = CacheBuilder.newBuilder().maximumSize(100).weakValues().build();
+
+    static final Cache<FlowNode,String> execNodeNameCache = CacheBuilder.newBuilder().maximumSize(1000).expireAfterAccess(1, TimeUnit.HOURS).build();
+
     private static final Logger LOGGER = Logger.getLogger(FlowNodeUtil.class.getName());
 
     private FlowNodeUtil() {
     }
 
-    public static long getNodeExecDuration(FlowNode node) {
-        List<FlowNode> childNodes = getChildNodes(node);
-        if (!childNodes.isEmpty()) {
-            long startTime = TimingAction.getStartTime(node);
-            long endTime = TimingAction.getStartTime(childNodes.get(childNodes.size() - 1));
+    public static boolean isNotPartOfRunningBuild(FlowExecution execution) {
+        Queue.Executable executable = null;
```

Comment:
@jglick  I do *not* consider this overall ready for review, but wonder if this test makes sense?  It's borrowed from existing logic, but I'm not 100% sure if it is correct (I'm seeing some odd behaviors).  Is there a better way?  Perhaps something like the following?

```java
return (execution instanceof CpsFlowExecution && ((CpsFlowExecution)execution).isComplete());
```